src/NServiceBus.Core/Persistence/InMemory/Outbox/InMemoryOutboxStorage.cs

```diff
@@ -48,24 +47,21 @@ public Task SetAsDispatched(string messageId, ContextBag context)
                 return TaskEx.CompletedTask;
             }
 
-            storedMessage.TransportOperations.Clear();
-            storedMessage.Dispatched = true;
-
+            storedMessage.MarkAsDispatched();
             return TaskEx.CompletedTask;
         }
 
         public void RemoveEntriesOlderThan(DateTime dateTime)
         {
-            var entriesToRemove = storage
-                .Where(e => e.Value.Dispatched && e.Value.StoredAt < dateTime)
-                .Select(e => e.Key)
-                .ToList();
-
-            foreach (var entry in entriesToRemove)
+            foreach (var entry in storage.ToArray())
```

Comment:
:+1: wasn't entirely sure. Thx for claryfying

On Wednesday, July 6, 2016, Brandon Ording <notifications@github.com> wrote:

> In
> src/NServiceBus.Core/Persistence/InMemory/Outbox/InMemoryOutboxStorage.cs
> <https://github.com/Particular/NServiceBus/pull/3903#discussion_r69659933>
> :
>
> >              return TaskEx.CompletedTask;
> >          }
> >
> >          public void RemoveEntriesOlderThan(DateTime dateTime)
> >          {
> > -            var entriesToRemove = storage
> > -                .Where(e => e.Value.Dispatched && e.Value.StoredAt < dateTime)
> > -                .Select(e => e.Key)
> > -                .ToList();
> > -
> > -            foreach (var entry in entriesToRemove)
> > +            foreach (var entry in storage.ToArray())
>
> That's safe to do with ConcurrentDictionary.
>
> https://msdn.microsoft.com/en-us/library/dd287131(v=vs.110).aspx
>
> And since this is trying to purge old values, it shouldn't matter that
> it's not a moment-in-time snapshot either.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Particular/NServiceBus/pull/3903/files/9b3324353ab994bdf2836ce71186e54aed8cdac4#r69659933>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAKosnRcHNY-wClwu9lYCncJQWBxuAhGks5qSvVagaJpZM4JFK20>
> .
>